### PR TITLE
Add edits in locked state in split payment

### LINKED ISSUE #3791: Expenditures: Split payment - add edit option to locked state
[Provided by Adrian via Contractors project management platform]

## Description
User has to be able to edit split payment.

## Acceptance Criteria

- the ability to change the type of split i.e. equal/unequal
- the ability to change and remove the recipients in both equal/unequal split
- the ability to change the percent of split for each recipient when it is an unequal split
- the ability to change the value of the split and the token selected

## Design

- [Figma Link - Split payments edit state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=6179%3A333651)

Screenshot 2022-09-01 100701

### PR DESCRIPTION
**Pull request by agnieszkajarosikloj**
## Description

This PR adds  edit option to locked state in Split payment. Edition is enabled after clicking 'Edit' icon. 
When user add some changes and click on 'Confirm' button, he sees changes listed in modal. Changes require confirmation.
This PR adds changes only in Split payment type. Changes in other payments will be added in next PRs.

## Acceptance Criteria

- the ability to change the type of split i.e. equal/unequal
- the ability to change and remove the recipients in both equal/unequal split
- the ability to change the percent of split for each recipient when it is an unequal split
- the ability to change the value of the split and the token selected

Screenshot 2022-10-05 at 09 10 00

Resolves #3791

### CODE REVIEW

**Review by willm30 — CHANGES_REQUESTED** [review]
Looking good to me! Just noticing a couple of things:

1. The discard button remains after discarding the change to the split when there are multiple sections (e.g. change title also):
discard-remains

2. The cursor for the edit icon should be `cursor: pointer`

**Comment by agnieszkajarosikloj** [comment]
@willm30 I have added fix. Can you have a look please?

**Review by willm30 — COMMENTED** [review]
Thanks. Just a couple of questions, the same that I left on one of the other edit prs. 

Would it make more sense to have it "Change recipient: 1", since the other headings have "change" in them? And changing "Change to" to just "To" to avoid repeating "Change"? 

And also noticing here that the user's name doesn't have the "@" in front, like in the design.

**Comment by agnieszkajarosikloj** [comment]
@willm30 I have added changes. Can you have a look please?

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import { ValueOf } from '../../ChangedValues/ChangedValues';
+
+import { skip } from '../ChangedSplit';
+
+import styles from './ChangedRecipient.css';
+
+export const MSG = defineMessages({
+  removed: {
+    id: `dashboard.EditExpenditureDialog.ChangedStaged.ChangedRecipient.removed`,
+    defaultMessage: 'Mielstone has been deleted',
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx:19]: Typo: "Mielstone"

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import ChangeItem from '../../ChangedMultiple/ChangeItem';
+import { ValueOf } from '../../ChangedValues/ChangedValues';
+
+import { skip } from '../ChangedSplit';
+
+import styles from './ChangedRecipient.css';
+
+export const MSG = defineMessages({
+  removed: {
+    id: `dashboard.EditExpenditureDialog.ChangedStaged.ChangedRecipient.removed`,
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx:18]: Id should be ChangedSplit

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    defaultMessage: 'Mielstone has been deleted',
+  },
+});
+
+const displayName = `dashboard.EditExpenditureDialog.ChangedStaged.ChangedRecipient`;
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx:23]: Same

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        ([recipientKey, recipientValue], index) => {
+          // milestoneKey - 'user', 'amount', 'id', 'removed'
+          if (skip.includes(recipientKey)) {
+            return null;
+          }
+
+          if (recipientKey === 'removed') {
+            return (
+              // eslint-disable-next-line react/no-array-index-key
+              <div className={styles.row} key={index}>
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx:44]: Reason for using index as key as opposed to `nanoId`?

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+              </div>
+            );
+          }
+
+          return (
+            <ChangeItem
+              newValue={recipientValue as ValueOf<ValuesType>}
+              oldValue={oldRecipient?.[recipientKey]}
+              // eslint-disable-next-line react/no-array-index-key
+              key={index}
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedRecipient/ChangedRecipient.tsx:60]: Same q re index as key

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedSplit.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    | ValuesType['recipients']
+    | string
+    | Staged
+    | Pick<
+        LoggedInUser,
+        'walletAddress' | 'balance' | 'username' | 'ethereal' | 'networkId'
+      >
+    | Batch;
+};
+
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedSplit.tsx:59]: This is defined in Changed Staged. Only need one.

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedSplit.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  return (
+    <>
+      {changedSplit.map(([key, value], index) => {
+        const oldValue = oldValues[newValues.key]?.[key];
+
+        return (
+          // eslint-disable-next-line react/no-array-index-key
+          <Fragment key={index}>
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedSplit/ChangedSplit.tsx:106]: Index as key here?

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialog.tsx [hunk 8]
```diff
@@ -78,6 +78,23 @@ const EditExpenditureDialog = ({
...
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const discardPaymentChange = useCallback(
+    (payment: ExpenditureTypes, name: keyof ValuesType) => {
+      setConfirmedValues((confirmedVal) => {
+        if (confirmedVal && !(name in confirmedVal[payment])) {
+          return undefined;
+        }
+        const newPayment = confirmedVal && confirmedVal[payment];
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const { [name]: removedProperty, ...updatedValues } = newPayment || {};
```
Comment (willm30) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/EditExpenditureDialog.tsx:90]: Reason for unused var?

File: src/modules/dashboard/components/ExpenditurePage/Split/SplitEqual/SplitEqual.tsx [hunk 9]
```diff
@@ -91,6 +92,31 @@ const SplitEqual = ({ colony, sidebarRef }: Props) => {
...
   }, [amount, recipientsCount]);
 
+  useEffect(() => {
+    setValue(
+      recipients.map((recipient) => ({
+        ...recipient,
+        amount: { value: calculatedAmount, tokenAddress: amount.tokenAddress },
+      })),
+    );
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (willm30) [inline src/modules/dashboard/components/ExpenditurePage/Split/SplitEqual/SplitEqual.tsx:102]: Perhaps leave a comment here that setValue causes infinite loop?

File: src/modules/dashboard/components/ExpenditurePage/Split/SplitEqual/SplitEqual.tsx [hunk 10]
```diff
@@ -91,6 +92,31 @@ const SplitEqual = ({ colony, sidebarRef }: Props) => {
...
+        recipients.map((rec) => ({
+          ...rec,
+          amount: {
+            value: rec.amount?.value,
+            tokenAddress: value,
+          },
+        })),
+      );
+    },
+    [recipients, setValue],
```
Comment (willm30) [inline src/modules/dashboard/components/ExpenditurePage/Split/SplitEqual/SplitEqual.tsx:117]: Given that setValue is recreated on every render, there's no point in using `useCallback` here -- or, remove from dependencies array.

File: src/modules/dashboard/components/ExpenditurePage/Split/SplitUnequal/SplitUnequal.tsx [hunk 11]
```diff
@@ -77,28 +78,63 @@ const SplitUnequal = ({ colony, sidebarRef }: Props) => {
...
+  }, [recipients]);
+
+  const onSliderChange = useCallback(
+    (val: number, name: string) => {
+      setFieldValue(name, {
+        value: (Number(val) / 100) * Number(amount.value),
+        tokenAddress: amount.tokenAddress,
+      });
+    },
+    [amount.value, setFieldValue, amount.tokenAddress],
```
Comment (willm30) [inline src/modules/dashboard/components/ExpenditurePage/Split/SplitUnequal/SplitUnequal.tsx:105]: I think same will apply to setFieldValue, i.e. recreated each render

File: src/modules/dashboard/components/ExpenditurePage/Split/SplitUnequal/SplitUnequal.tsx [hunk 12]
```diff
@@ -77,28 +78,63 @@ const SplitUnequal = ({ colony, sidebarRef }: Props) => {
...
+        recipients.map((rec) => ({
+          ...rec,
+          amount: {
+            value: (Number(rec.percent) / 100) * Number(value),
+            tokenAddress: amount.tokenAddress,
+          },
+        })),
+      );
+    },
+    [recipients, setValue, amount.tokenAddress],
```
Comment (willm30) [inline src/modules/dashboard/components/ExpenditurePage/Split/SplitUnequal/SplitUnequal.tsx:136]: Same thought re setValue and useCallback here and above

**Review by willm30 — CHANGES_REQUESTED** [review]
Minor comments on the code but this looks good to me so can re-review once you've had a look through.

**Comment by agnieszkajarosikloj** [comment]
@willm30 I have added changes. It's ready for re-review.

**Review by willm30 — APPROVED** [review]
Nice one. Leaving the same comments here as the other one but looks good to me these aside.

1. The user avatar and token avatars aren't same size
2. Could we just have "Change Amount" instead of "Change Split Amount"? 
3. Could we capitalise "title" and "description" in "Change Title" and "Change Description"?

**Comment by agnieszkajarosikloj** [comment]
@JoinColony/frontend can you review it, please?

**Review by chinins — APPROVED** [review]
Well done - quite a PR!

**Pull request merged**